### DISCUSSION TITLE
Improve Selection/Cropping mechanics

### DISCUSSION THREAD

**Discussion by pureby** [sylikc/jpegview#151] [Ideas]
1. Add **1:1** aspect ratio to the list of default "Crop modes" (currently, 1:1 aspect ratio is only available from `UserCropAspectRatio=1 1`, making it impossible to add a custom aspect ratio _and_ keep 1:1 option).

2. Add **vertical** aspect ratios (there is a 16:9 option, but no 9:16).

3. Make `UserCropAspectRatio` accept **multiple values** (e.g. `UserCropAspectRatio=1 2,1 3,1 4`)

4. _**Even better:** move all aspect ratios to `CropAspectRatios=1 1,5 4,4 3,3 2,16 9,16 10`, so users can add and remove aspect ratios._

5. Add `DefaultCropAspectRatio` setting (of course, _Free_ mode should remain the default, but it would be nice to be able to override this behavior (I use 1:1 crop mode nine out of ten times, for example)).

**Comment by sylikc** [comment]
1. I'm actually doing the crop AR thing right now so... I could add that
2. this one's a bit tricky... wouldn't want to overwhelm the list, and I guess I'd prefer how other programs do it, where if you start dragging too much in one direction, it flips... but it's not quite the way it was written right now.  I'm in the code, so I'll see if I can change this...
3. Right now UserCropAspectRatio can be bound the a hotkey... if doing it this way, you can't
4. Same thing, the other ratios can actually be bound to a hotkey (even though this isn't exposed right now) ... moving it to like UCAR0=, UCAR1= would be a pretty involved addition
5. Can be considered, once again, interaction with the keymap mechanism ..

**Reply by sylikc** [reply]
@pureby I just pushed a commit that adds a whole bunch of these features... can you test it for me?  I think it exhibits the behavior I would like, but since it's such a big change, I need a bit of feedback.

1. I added the 1:1 ratio
2. There is a new feature when you drag the topleft or bottomright which will invert to a vertical ratio when dragging way out of bounds... it's like other programs, so without me telling you how, try it and see if it flows the way you'd expect

**Reply by sylikc** [reply]
Do you know how to download the latest build?

**Reply by pureby** [reply]
Awesome! Sure, I want to test. And no, I don't know where to get the test build.

**Reply by sylikc** [reply]
~~Click on the latest successful run, and find at the bottom a release build:~~

I created a new wiki page... let me know if it needs to be cleaned up: https://github.com/sylikc/jpegview/wiki/Dev-Snapshots#links

**Reply by pureby** [reply]
Tested `JPEGView-GH_sylikc-Release_x64-149a39e.zip`

1. It would be more logical to organize crop aspect ratios by their shape, starting with square and gradually going towards the most oblong ratio, so: `1:1, 5:4, 4:3, 3:2, 16:10, 16:9` (place 1:1 as first and 16:10 before 16:9). I understand this will probably conflict with existing keyboard shortcuts, but IMHO, the consistency of the UI is more important. And since you're doing a big change, why not do it right even if it breaks a couple old workflows temporarily.

2. "Same as image" — it's very nice to see that it actually displays image AR!

3. Flipping AR from horizontal to vertical: I think you nailed it. Works very well for me! The only issue is, if I didn't _know_ this was possible in advance, I would have never guessed how to do it. Even when I knew that I should be able to flip the AR by "dragging something outside of it's zone", nothing I tried worked until I came back your post and read that I must drag the TL or BR corner.

4. Personally, I hate tips in apps, but maybe in this case a tip would be justified.
Adding a _"Switch to vertical selection"_ menu item under "Zoom to selection" in the selection pop-up menu might help. And let's say, if user has clicked that option at least twice, the third and fourth time he clicks that menu item display a short tip like: _"You can flip selection by dragging top-left or bottom-right corner"?_ I think this will be enough explanation, as for me, once you know what part to dag it flips very intuitively every time.
Also, I don't know of any other software that does this trick. I just tried it in GIMP and it didn't work (but I have a strong feeling GIMP was never intended for human use :-D).

**Reply by pureby** [reply]
> 5. _Add `DefaultCropAspectRatio` setting_

> 5. Can be considered, once again, interaction with the keymap mechanism ..

How about `RememberLastCropRatio=1` option as an alternative? That would be even better than `DefaultCropAspectRatio` and won't interfere with keymaps.

**Reply by sylikc** [reply]
Anything that has to do with "remembering" would require writing to the user ini.  I'm currently thinking of ideas as to how to indicate which INI an option is coming from or whatnot... in debugging it's messy as @#$!*(!

I'll keep these suggestions in mind though, just backlog for now holding out on figuring out the interaction of the inis

**Reply by pureby** [reply]
Ah, I see. I thought this would be easy because `DefaultWindowRect=sticky` does something similar, but I can understand that this is a mess for debugging when adding new settings.
